### PR TITLE
Issue-702 Imrpovement of `cloudProviderSettings` flow

### LINKED ISSUE #702: Categorize `CloudProviderSettings`
In response to the InstAPI v2 documentation update and cloud provider settings expansion, we are considering restructuring our cloud settings. Instead of retaining all configurations within a single struct like `CloudProviderSettings`, I propose dividing its fields into appropriate groups. For example, we could organize settings related to Azure into a substructure named `azureSettings`, including fields such as `resourceGroup, customVirtualNetworkId, and storageNetwork`. This approach aims to enhance clarity and maintainability in our configuration structure, aligning it with the updated documentation.

### PR DESCRIPTION
**Pull request by worryg0d**
The PR provides devideds the whole `cloudProviderSettings` into different object for each cloud provider.

Closes #702

### CODE REVIEW

File: apis/clusters/v1beta1/structs.go [hunk 1]
```diff
@@ -820,3 +820,110 @@ func (g GenericResizeSettings) Equal(o GenericResizeSettings) bool {
...
+			DisableSnapshotAutoExpiry: instaModel.GCPSettings[0].DisableSnapshotAutoExpiry,
+		}
+	case len(instaModel.AzureSettings) > 0:
+		s.AzureSettings = &AzureSettings{
+			ResourceGroup:          instaModel.AzureSettings[0].ResourceGroup,
+			CustomVirtualNetworkID: instaModel.AzureSettings[0].CustomVirtualNetworkID,
+			StorageNetwork:         instaModel.AzureSettings[0].StorageNetwork,
+		}
+	}
+}
```
Comment (testisnullus) [inline apis/clusters/v1beta1/structs.go:872]: Here we will append only if `s` is not nil:
```
func (c *CloudProviderOptions) FromInstAPI(instaModel *models.CloudProviderSettings) {
	var s *CloudProviderOpts

	switch {
	case len(instaModel.AWSSettings) > 0:
		s = &CloudProviderOpts{
			AWSSettings: &AWSSettings{
				EncryptionKey:          instaModel.AWSSettings[0].EBSEncryptionKey,
				CustomVirtualNetworkID: instaModel.AWSSettings[0].CustomVirtualNetworkID,
				BackupBucket:           instaModel.AWSSettings[0].BackupBucket,
			},
		}
	case len(instaModel.GCPSettings) > 0:
		s = &CloudProviderOpts{
			GCPSettings: &GCPSettings{
				CustomVirtualNetworkID:    instaModel.GCPSettings[0].CustomVirtualNetworkID,
				DisableSnapshotAutoExpiry: instaModel.GCPSettings[0].DisableSnapshotAutoExpiry,
			},
		}
	case len(instaModel.AzureSettings) > 0:
		s = &CloudProviderOpts{
			AzureSettings: &AzureSettings{
				ResourceGroup:          instaModel.AzureSettings[0].ResourceGroup,
				CustomVirtualNetworkID: instaModel.AzureSettings[0].CustomVirtualNetworkID,
				StorageNetwork:         instaModel.AzureSettings[0].StorageNetwork,
			},
		}

		if s != nil {
			*c = append(*c, s)
		}
	}
}
```

wdyt?

Comment (worryg0d) [reply apis/clusters/v1beta1/structs.go]: It makes sense. Done.

File: Makefile [hunk 2]
```diff
@@ -142,8 +142,8 @@ uninstall: manifests kustomize ## Uninstall CRDs from the K8s cluster specified
 deploy: manifests kustomize ## Deploy controller to the K8s cluster specified in ~/.kube/config.
 	cd scripts && ./make_creds_secret.sh
 	cd config/manager && $(KUSTOMIZE) edit set image controller=${IMG}
-	$(KUSTOMIZE) build config/default | kubectl apply -f -
 	kubectl apply -f ~/creds_secret.yaml
+	$(KUSTOMIZE) build config/default | kubectl apply -f -
```
Comment (ribaraka) [inline Makefile:146]: pls comment kustomization

File: apis/clusters/v1beta1/generic_spec.go [hunk 3]
```diff
@@ -99,14 +99,40 @@ func (s *GenericClusterSpec) ClusterSettingsUpdateToInstAPI() *models.ClusterSet
...
+	// For customers provisioning on Instaclustr's cloud provider accounts, this property may be omitted.
+	//
 	//+kubebuilder:default:=INSTACLUSTR
-	ProviderAccountName   string                   `json:"accountName,omitempty"`
-	Network               string                   `json:"network"`
-	Tags                  map[string]string        `json:"tags,omitempty"`
-	CloudProviderSettings []*CloudProviderSettings `json:"cloudProviderSettings,omitempty"`
+	ProviderAccountName string            `json:"accountName,omitempty"`
+	Network             string            `json:"network"`
+	Tags                map[string]string `json:"tags,omitempty"`
```
Comment (ribaraka) [inline apis/clusters/v1beta1/generic_spec.go:120]: doc this

File: apis/clusters/v1beta1/generic_spec.go [hunk 4]
```diff
@@ -158,6 +165,64 @@ func (s *GenericDataCentreSpec) ToInstAPI() models.GenericDataCentreFields {
...
+
+func (s *GenericDataCentreSpec) cloudProviderSettingsFromInstAPI(instaModel *models.CloudProviderSettings) {
+	if instaModel == nil {
+		return
+	}
+
+	switch {
+	case len(instaModel.AWSSettings) > 0:
+		setting := instaModel.AWSSettings[0]
+		s.AWSSettings = append(s.AWSSettings, &AWSSettings{
```
Comment (ribaraka) [inline apis/clusters/v1beta1/generic_spec.go:209]: don't use append

File: apis/clusters/v1beta1/validation.go [hunk 5]
```diff
@@ -432,10 +427,42 @@ func (s *GenericDataCentreSpec) ValidateOnPremisesCreation() error {
...
+	}
+
+	if fieldSet > 1 {
+		return errors.New("only one of [awsSettings, gcpSettings, azureSettings] should be set")
+	}
+
+	return nil
+}
+
+func (s *GenericDataCentreSpec) cloudProviderSettingsPresented() bool {
```
Comment (ribaraka) [inline apis/clusters/v1beta1/validation.go:466]: hasCloudProviderSettings()

File: apis/clusters/v1beta1/validation.go [hunk 6]
```diff
@@ -432,10 +427,42 @@ func (s *GenericDataCentreSpec) ValidateOnPremisesCreation() error {
...
+	if s.AzureSettings != nil {
+		fieldSet++
+	}
+	if s.GCPSettings != nil {
+		fieldSet++
+	}
+
+	if fieldSet > 1 {
+		return errors.New("only one of [awsSettings, gcpSettings, azureSettings] should be set")
+	}
```
Comment (ribaraka) [inline apis/clusters/v1beta1/validation.go:461]:
```go
if sum := len(s.AWSSettings) + len(s.AzureSettings) + len(s.GCPSettings); sum > 1 { return err } 
```

**Review by ribaraka — CHANGES_REQUESTED** [review]

File: apis/clusters/v1beta1/generic_spec.go (hunk 4 above)
Comment (worryg0d) [reply apis/clusters/v1beta1/generic_spec.go]: Done

File: apis/clusters/v1beta1/validation.go (hunk 5 above)
Comment (worryg0d) [reply apis/clusters/v1beta1/validation.go]: Done

**Review by ribaraka — APPROVED** [review]

**Pull request merged**